packages/dds/tree/src/shared-tree/treeCheckout.ts

```diff
@@ -420,11 +438,23 @@ export class TreeCheckout implements ITreeCheckoutFork {
 				this.events.emit("afterBatch");
 			}
 			if (event.type === "replace" && getChangeReplaceType(event) === "transactionCommit") {
+				// commit the in progress removed roots if there are no more transactions
+				if (!this.transaction.inProgress()) {
+					assert(
+						this.removedRootsInTransaction !== undefined,
+						"a transaction should operate on a copy of the removed roots index",
+					);
+					this.removedRoots.loadIndex(this.removedRootsInTransaction);
+					this.removedRootsInTransaction = undefined;
+				}
 				const transactionRevision = event.newCommits[0].revision;
 				for (const transactionStep of event.removedCommits) {
 					this.removedRoots.updateMajor(transactionStep.revision, transactionRevision);
 				}
 			}
+			if (event.type === "remove") {
+				this.removedRootsInTransaction = undefined;
+			}
```

Comment:
This will purge the removed roots when an inner transaction is aborted. You at least need to check if we're still in a transaction:
```suggestion
			if (event.type === "remove" && !this.transaction.inProgress()) {
				this.removedRootsInTransaction = undefined;
			}
```
Unfortunately this is also not quite right because it now means that we're not reverting to the point before the subtransaction. I think we're going to need a stack of snapshots to match the stack of (sub)transactions.